Reintroduce image scanning for openSUSE and SLE

flavio opened the pull request:
As promised inside of https://github.com/coreos/clair/pull/503... handle scanning of openSUSE and SUSE Linux Enterprise images.

The unit tests are passing but I wasn't able to test the code against real images because neither [clair-scanner](https://github.com/arminc/clair-scanner) nor [reg](https://github.com/jessfraz/reg) work against `master`'s new API. Suggestions about how to test the changes are welcome 😃

flavio committed:
Reintroduce image scanning for openSUSE and SLE

Handle scanning of openSUSE and SUSE Linux Enterprise images.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

jzelinskie commented:
Thank you so much for reworking all of these changes! I'm really sorry for the slow response.

We don't really have an e2e test framework, and the API changed this release. You can do manual testing by simply using cURL against the API server. It handles both gRPC and JSON over REST.

I'd also like to see some changes to the documentation to include the licensing information for the data in the openSUSE database.

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+
+	log "github.com/sirupsen/logrus"
+
+	"fmt"
```
this should be grouped in the first section of imports

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+		up.UpdaterFlag = "openSUSEUpdater"
+		up.FileRegexp = regexp.MustCompile(`opensuse.leap.(\d+\.*\d*).xml`)
+	default:
+		panic("Unrecognized flavor")
```
panic("tried to create an updater for an unrecognized flavor of suse")

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+	if err != nil {
+		return resp, err
+	}
+	log.WithField("flagvalue", flagValue)
```
This probably needs to be something like Debug level if it stays at all.

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+
+	// this contains the modification time of the most recent
+	// file expressed as unix time (int64)
+	latestOval, _ := strconv.ParseInt(flagValue, 10, 64)
```
Why ignore this error?

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+			log.Fields{
+				"ovalFile": ovalFile,
+				"updater":  u.Name,
+			}).Debug("file to check")
```
```go
	log.WithFields(log.Fields{
		"ovalFile": ovalFile,
		"updater":  u.Name,
	}).Debug("file to check")
```

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+				"updater":  u.Name,
+			}).Debug("file to check")
+
+		// do not fetch the entire file to get the value of the
```
Do

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+	}
+
+	// "Thu, 30 Nov 2017 03:07:57 GMT
+	layout := "Mon, 2 Jan 2006 15:04:05 MST"
```
pull this out into a constant `timeFormatLastModified`

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+		if err != nil {
+			log.WithError(err).Error("could not download", u.Name, "update list")
+			return resp, commonerr.ErrCouldNotDownload
+		}
```
`defer r.Body.Close()` or make sure to put `r.Body.Close()` before any return/continue statements

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+	}
+
+	// timestamp format 2017-10-23T04:07:14
+	layout := "2006-1-2T15:04:05"
```
pull this out into a constant `timeFormatOVAL`

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+				Vulnerability: database.Vulnerability{
+					Name: name(definition),
+					Link: link(definition),
+					//TODO: handle that once openSUSE/SLE OVAL files have severity info
```
pull this out into a function `severity(definition)` that doesn't do anything but return UnknownSeverity, and has this comment.

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+		// Attempt to parse package data from trees of criterions.
+		for _, c := range criterions {
+			if match := suseInstalledCommentRegexp.FindStringSubmatch(c.Comment); match != nil {
+				if len(match) != 4 {
```
it's easier to read this conditional if you keep it `if (positive case) else negative case`.

```go
if len(match) == 4 {
  osVerison = ...
} else {
  log.WithField().Warning()
}
```

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,473 @@
+	re := regexp.MustCompile(`-\d+\.`)
+
+	matches := re.FindStringSubmatchIndex(fullname)
+
```
get rid of this new line

jzelinskie commented on ext/vulnsrc/suse/suse_test.go:
add some tests for the failure scenario

jzelinskie reviewed (changes requested):
This mostly looks good since it's a lot of copypasta from the the RHEL vulnsrc. Ideally after this is merged, we can refactor some of the OVAL parsing into a library decoupled from the particular distribution.

flavio commented:
@jzelinskie sorry for the late response... I've implemented all the changes you requested. Once you are fine with the changes I'll rebase this branch against master and squash all the commits into a single one.

:heart:

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,481 @@
+	// Fetch the update list.
+	r, err := http.Get(ovalURI)
+	if err != nil {
+		log.Fatal(err)
```
This shouldn't be fatal.

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,481 @@
+		// Collect vulnerabilities.
+		for _, v := range vs {
+			resp.Vulnerabilities = append(resp.Vulnerabilities, v)
+		}
```
This for-loop can be done in one line:

```go
resp.Vulnerabilities = append(resp.Vulnerabilities, vs...)
```

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,481 @@
+			}
+			for _, p := range pkgs {
+				vulnerability.Affected = append(vulnerability.Affected, p)
+			}
```
This for-loop can be done in one line:

```go
vulnerability.Affected = append(vulnerability.Affected, pkgs...)
```

jzelinskie commented on ext/vulnsrc/suse/suse.go:
```diff
@@ -0,0 +1,481 @@
+			}
+
+			if suseOpenSUSEInstalledCommentRegexp.FindStringSubmatch(c.Comment) == nil && strings.HasSuffix(c.Comment, " is installed") {
+				name, version, err := splitPackageNameAndVersion(c.Comment[:len(c.Comment)-13])
```
In order to make it obvious where `13` is coming from: `len(c.Comment)-len(" is installed")`

jzelinskie reviewed (changes requested):
couple of tiny nitpicks, but feel free to rebase now

flavio commented:
@jzelinskie I've update the code with the changes you proposed. Take a look at the changes I made and gimme the final green light for the rebase.

jzelinskie commented:
I don't want this to hang any longer. Rebase away!

flavio commented:
> I don't want this to hang any longer. Rebase away!

@jzelinskie done!

flavio commented:
I did another forced push to address the style check errors reported by `goftm`.

Payback159 commented:
Hi Guys, any forecast when this feature will be available?

jzelinskie commented:
Sorry for the silence around the holidays. This code all looks fine. I think all this PR needs is to add OpenSUSE to the documentation in a few places. Most importantly here: https://github.com/coreos/clair/blob/master/Documentation/drivers-and-data-sources.md#data-sources-for-the-built-in-drivers

flavio committed:
Update documentation: talk about SUSE support

Expand the documentation about the available data sources to mention
openSUSE and SLE.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
@jzelinskie I did a rebase against master and then added a new commit which extends the documentation. Thanks for pointing that out. AFAIK this is the only place of the documentation that needs to be extended.

jzelinskie reviewed (approved):
Yay! Let's get this ball rolling!
